File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [sys.maxsize] * (n + 1)
        dp[0] = 0

        for i in range(int(math.sqrt(n)) + 1):
            square_i = i * i
            for j in range(square_i, n + 1):
                dp[j] = min(dp[j], dp[j - square_i] + 1)
        
        return dp[n]

File: 0279-perfect-squares/0279-perfect-squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function(n) {
    let dp = Array(n + 1).fill(Number.MAX_SAFE_INTEGER);
    dp[0] = 0;
    
    for (let i = 1; i * i <= n; i++) {
        for (let j = 1; j <= n; j++) {
            if (i * i <= j) {
                dp[j] = Math.min(dp[j], dp[j - i * i] + 1);
            }
        }
    }
    
    return dp[n];
};